haystack/nodes/preprocessor/preprocessor_new.py

```diff
@@ -0,0 +1,475 @@
+
+        :param split_overlap: Units (for example words or sentences) overlap between two adjacent documents after a split.
+                                For example, if `split_by="word" and split_length=5 and split_overlap=2`, then the splits would be like:
+                                `[w1 w2 w3 w4 w5, w4 w5 w6 w7 w8, w7 w8 w10 w11 w12]`.
+                                Set the value to 0 to ensure there is no overlap among the documents after splitting.
+
+        :param split_max_chars: Absolute maximum number of chars allowed in a single document. Reaching this boundary
+                            cut the document, even mid-word, and log a loud error.\n
+                            It's recommended to set this value to approximately double the size you expect your documents
+                            to be. For example, with `split_by='sentence'`, `split_lenght=2`, if the average sentence
+                            length of our document is 100 chars, you should set `max_char=400` or `max_char=500`.\n
+                            This is a safety parameter to avoid extremely long documents to end up in the document store.
+                            Keep in mind that huge documents (tens of thousands of chars) will strongly impact the
+                            performance of Reader nodes and can drastically slow down the indexing speed.
+
+        :param split_max_tokens:  Maximum number of tokens that are allowed in a single split. If set to 0, it will be
+                            ignored. If set to any value above 0, it requires `tokenizer_model` to be set to the
+                            model of your Reader and will verify that, whatever your `split_length` value is set
+                            to, the number of tokens included in the split documents will never be above the
+                            `max_tokens` value. For example:
+
+                            ```python
+                            PreProcessor(split_by='sentence', split_length=10, max_tokens=512, max_chars=5000, ...)
+                            ```
+
+                            means:
+
+                            - Documents will contain whole sentences
+                            - Documents will contain at most 10 sentences
+                            - Documents might contain less than 10 sentences if the maximum number of tokens is
+                                reached earlier.
+                            - Documents will never contain more than 5000 chars. Documents with a content length
+                                above that value will be split on the 5000th character.
+
+                            Note that the number of tokens might still be above the maximum if a single sentence
+                            contains more than 512 tokens. In this case an `ERROR` log is emitted, but the document
+                            is generated with whatever amount of tokens the first sentence has.
```

Comment:
Thank you for the feedback! I'm also curious what @robpasternak will say on this topic. For now I got no direct opinions on this detail yet.

Reply:
Sorry for the delayed reply! I've put the question to my fellow Sols. While waiting to hear back from them, here are my own questions and opinions:
1. What are folks' thoughts about the possibility of a separate bool parameter to determine whether `split_max_tokens` is a hard or soft cap? I could see the flexibility being useful, but it's also easy for me to suggest new parameters from the sidelines. Perhaps this new param might be odd to do specifically with `split_max_tokens`, but then again token counts are obviously particularly salient.
2. For what it's worth I think this issue isn't _that_ rare. I ran `NewPreProcessor` with the jazz data (344 wiki pages) and `split_max_tokens = 512`, and of the ~2700 resulting docs 28 of them were longer than `split_max_tokens` (max was 1490). And I've seen real-world cases where several docs were thousands of words longer than the split length (even with `split_respect_sentence_boundary = False`) because of how sentences were parsed.
3. If a sentence is cut off because of a hard `split_max_tokens` cap, what will the next doc look like? Will it just start where the previous one left off (modulo any `split_overlap`)?

Reply:
So after talking with other Sols it looks like there's definitely a desire for a `split_max_tokens` hard cap. I get the sense that there isn't much desire for a soft-capped version, so perhaps @tholor 's suggestion instead of my make-hard-cap-optional suggestion is the right way to go.

Reply:
Hey, @robpasternak thanks for the analysis. I would like to ask some follow-up questions on one of your observations:

**For Number 2**:
For the 28 Documents that had sentences longer than 512 tokens in the Jazz dataset, were those actual sentences or very long text snippets due to a lack of punctuation? For example, long bullet point lists will often not be detected as individual sentences since they aren't grammatically written as complete sentences.

**For this case:**

> And I've seen real-world cases where several docs were thousands of words longer than the split length (even with split_respect_sentence_boundary = False) because of how sentences were parsed.

What were the options that you used in the previous `PreProcessor` that resulted in this error? And would this be a good integration/real-world test to add to the PR to test the `NewPreProcessor`?